packages/blocks/src/__internal__/service/json2block.ts

```diff
@@ -28,131 +28,178 @@ export async function json2block(
-      if (model.text) {
-        await setRange(model, {
-          index: textLength,
-          length: 0,
-        });
-      }
-    }
 
-    isFocusedBlockEmpty && page.deleteBlock(focusedBlockModel);
-    return;
+  let lastItemMergedIds: string[] = [];
+  let splitSuffixModel: BaseBlockModel | null = null;
+  /**
+   * we should record focused block's children before change,
+   * it will move to correct position later
+   */
+  const focusedBlockChildrenBeforeChange = [...focusedBlockModel.children];
+
+  /**
+   * check if split need, it only should split on the caret between in focused block,
+   * and there are multiple pasted blocks or single nested pasted block
+   */
+  if (
+    !isFocusedBlockEmpty &&
+    textRangePoint.index > 0 &&
+    Number(focusedBlockModel.text?.length) > textRangePoint.index &&
+    (!isSinglePastedBlock || firstBlock.children.length > 0)
+  ) {
+    await handleBlockSplit(page, focusedBlockModel, textRangePoint.index, 0);
+    splitSuffixModel = page.getNextSibling(focusedBlockModel);
   }
 
-  await handleBlockSplit(page, focusedBlockModel, textRangePoint.index, 0);
-
   if (shouldMergeFirstBlock) {
     focusedBlockModel.text?.insertList(
       firstBlock.text || [],
       textRangePoint.index || 0
     );
-    await addSerializedBlocks(
+
+    lastItemMergedIds = await addSerializedBlocks(
       page,
       firstBlock.children || [],
       focusedBlockModel,
       0
     );
   }
 
-  const insertPosition =
-    parent.children.indexOf(focusedBlockModel) +
-    (shouldMergeFirstBlock ? 1 : 0);
-
-  const ids = await addSerializedBlocks(
-    page,
-    pastedBlocks.slice(shouldMergeFirstBlock ? 1 : 0),
-    parent,
-    insertPosition
+  const remainingPastedBlocks = pastedBlocks.slice(
+    shouldMergeFirstBlock ? 1 : 0
   );
 
-  isFocusedBlockEmpty && page.deleteBlock(focusedBlockModel);
+  if (remainingPastedBlocks.length > 0) {
+    const insertPosition =
+      parent.children.indexOf(focusedBlockModel) +
+      (shouldMergeFirstBlock ? 1 : 0);
 
-  const lastModel = page.getBlockById(ids[ids.length - 1]);
-  assertExists(lastModel);
+    lastItemMergedIds = await addSerializedBlocks(
+      page,
+      remainingPastedBlocks,
+      parent,
+      insertPosition
+    );
 
-  if (shouldMergeLastBlock) {
-    const rangeOffset = lastModel.text?.length || 0;
-    const nextSiblingModel = page.getNextSibling(lastModel);
-    lastModel.text?.join(nextSiblingModel?.text as Text);
-    assertExists(nextSiblingModel);
-    page.deleteBlock(nextSiblingModel);
-    // Wait for the block's rich text mounted
-    requestAnimationFrame(() => {
-      setRange(lastModel, {
-        index: rangeOffset,
-        length: 0,
-      });
-    });
-  } else {
-    if (lastModel?.text) {
-      setRange(lastModel, {
-        index: lastModel.text.length,
-        length: 0,
+    const lastBlock = pastedBlocks.at(-1);
+    if (lastBlock?.flavour === 'affine:image' && isFocusedBlockEmpty) {
```

Comment:
This has led to some use cases failing, causing them to assume that the editor content will only contain an image after an image is pasted.

But this will cause cursor lost. So I think there should be an empty block for focus or there already had a paragraph behind the image to focus after the paste.

If you all agree, I'll modify those test cases.

Reply:
This makes sense. Kindly proceed..